src/hypercube.c

```diff
@@ -178,7 +178,9 @@ ts_hypercube_from_constraints(ChunkConstraints *constraints, MemoryContext mctx)
 			DimensionSlice *slice;
 
 			Assert(hc->num_slices < constraints->num_dimension_constraints);
-			slice = ts_dimension_slice_scan_by_id(cc->fd.dimension_slice_id, mctx);
+			slice = ts_dimension_slice_scan_by_id_and_lock(cc->fd.dimension_slice_id,
```

Comment:
Why do we need to lock here, and why `SHARE` instead of `KEY SHARE`? I only see a need to lock on paths that mutate (adding or removing slices) and there's now protection in place for those paths.

At the very least, the choice of lock mode (and why locking is needed) warrants explanation.